tao_compiler/mlir/disc/transforms/fusion_utils.cc

```diff
@@ -486,69 +660,191 @@ void FusionPattern::calculateOperandsAndResults() {
-        for (Value rhs : op->getOperands().drop_front()) {
-          update(lhs, rhs, same_num_elements_impl_);
-          update(lhs, rhs, same_shape_impl_);
-        }
-      }
-      // Propagates same num_elements equality, not shape equality
-      if (isa<lmhlo::DynamicReshapeOp, lmhlo::ReshapeOp, lmhlo::TransposeOp>(
-              op)) {
-        Value input = op->getOperand(0);
-        // The last operand is the output memref by design
-        Value output = op->getOperand(num_operand - 1);
-        update(input, output, same_num_elements_impl_);
-      }
-      if (isa<lmhlo::ClampOp>(op)) {
-        Value min = op->getOperand(0);
-        Value operand = op->getOperand(1);
-        Value max = op->getOperand(2);
-        Value output = op->getOperand(3);
-        auto min_ty = min.getType().dyn_cast<MemRefType>();
-        auto operand_ty = operand.getType().dyn_cast<MemRefType>();
-        auto max_ty = max.getType().dyn_cast<MemRefType>();
-        if (!min_ty || !max_ty || !operand_ty) {
+// Create a new fusion pattern from the ops inside the lmhlo fusion op.
+FusionPattern::FusionPattern(lmhlo::FusionOp op, ShapeAnalysis* shape_analysis)
+    : FusionPatternBase(op) {
+  if (shape_analysis != nullptr) {
+    FusionType fusionType = FusionType::kNone;
+    auto deviceAttr = op->getAttrOfType<StringAttr>(kDiscPlaceAssignment);
+    auto fusionTypeAttr =
+        op->getAttrOfType<StringAttr>(kDiscFusionTypeAttrName);
+    if (fusionTypeAttr) {
+      fusionType = fusionTypeFromString(fusionTypeAttr.getValue());
+    }
+    if (!deviceAttr || fusionType == FusionType::kNone) {
+      fusion_type_ = FusionType::kNone;
+      dominant_op_ = (size() == 0 ? nullptr : getOpList()[0]);
+      return;
+    }
+    StringRef strategyStr = "base";
+    if (fusionType == FusionType::kStitch) {
+      strategyStr = "stitch";
+    }
+    FusionStrategy& strategy =
+        getFusionStrategy(deviceAttr.getValue(), strategyStr);
+    bool status = strategy.initFusionPattern(*shape_analysis, *this);
+    assert(status);
+    (void)(status);
+  }
+}
+
+// Create a new fusion pattern from a valid fusion op list.
+FusionPattern::FusionPattern(SmallVectorImpl<Operation*>& op_list)
+    : FusionPatternBase(op_list) {
+  // just set to a initialized state. It's responsibility of the user to
+  // reset the state.
+  fusion_type_ = FusionType::kNone;
+  dominant_op_ = (size() == 0 ? nullptr : getOpList()[0]);
+}
+
+// Merges two fusion patterns and returns the merged pattern. The original
+// pattern remains unmodified. The new merged pattern is uninitialized.
+FusionPattern FusionPattern::mergeWithoutInit(FusionPattern& other) {
+  FusionOpList new_op_list = getOpList();
+  new_op_list.insert(new_op_list.end(), other.getOpList().begin(),
+                     other.getOpList().end());
+  FusionPattern new_fusion_pattern{new_op_list};
+  return new_fusion_pattern;
+}
+
+bool FusionPattern::getOrderedSkeletonGroups(
```

Comment:
Do we need to put these gpu specific logics into an inherited class?

Reply:
I will make it a stand-alone function.